src/include/miopen/par_for.hpp

```diff
@@ -62,17 +62,16 @@ struct joinable_thread : std::thread
 struct thread_factory
 {
     template <class F>
-    joinable_thread operator()(std::size_t& work, std::size_t n, std::size_t grainsize, F f) const
+    joinable_thread operator()(std::size_t& work, std::size_t n, std::size_t threadsize, F f) const
```

Comment:
The grainsize should not be removed, as `n` can be greater than the `threadsize`.

Reply:
Different extension of kernel file has different compile time. In general, `cpp` > `.cl` > `.s`. I noticed that almost all solutions  returned by `FindDataImplicitGemmSolutions` contain the kernel files with extension `.cpp`, and some solutions  returned by `FindDataDirectSolutions` contain the kernel files with extension `.cl`, but almost all solutions return by `FindWinogradSolutions` contain the kernel files with extension `.s`. So, if I compile all solutions together, according to the original task allocation way, it may assign all kernel files with extension `.s` to a compile thread and all kernel files with extension `.cpp` to another compile thread. This will cause some threads to end quickly and some to run for a long time. In this case, performance may degrade. I modified this function to try to distribute tasks as evenly as possible.

Reply:
I see, you are doing a grid strided loop. This may make it more efficient for parallel compilation, but for other places `par_for` is used it will make it slower. It would be better to have a seperate function for this:

```cpp
template <class F>
void par_for_strided(std::size_t n, F f)
{
    const auto threadsize = std::min<std::size_t>(std::thread::hardware_concurrency(), n);
    par_for_impl(threadsize, threadsize, [&](auto start) {
        for(std::size_t i = start; i < n; i += threadsize)
        {
            f(i);
        }
    });
}
```